File: algorithmic_toolbox/week5_dynamic_programming_1/3_edit_distance/edit_distance.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int min(int a, int b) {
  return a < b ? a : b;
}

/* Levenshtein distance: insertion, deletion, substitution */
int edit_distance(const char* s, const char* t) {
  if (strcmp(s, t) == 0) return 0;

  int m = strlen(s), n = strlen(t);
  int dp[m + 1][n + 1];

  for (int i = 0; i <= m; ++i) dp[i][0] = i;

  for (int j = 0; j <= n; ++j) dp[0][j] = j;

  for (int i = 1; i <= m; ++i)
    for (int j = 1; j <= n; ++j)
      if (s[i - 1] == t[j - 1])
        dp[i][j] = dp[i - 1][j - 1];
      else
        dp[i][j] = min(dp[i - 1][j], min(dp[i][j - 1], dp[i - 1][j - 1])) + 1;

  return dp[m][n];
}

int main() {
  char s[1001], t[1001];
  int _ = scanf("%s %s", s, t);
  printf("%d\n", edit_distance(s, t));
  return 0;
}

File: algorithmic_toolbox/week5_dynamic_programming_1/3_edit_distance/edit_distance.py
from typing import Sequence


def edit_distance(s: Sequence, t: Sequence) -> int:  # noqa: C901
    """Levenshtein distance: insertion, deletion, substitution"""

    if s == t:
        return 0

    m, n = len(s), len(t)
    dp = [[0 for _ in range(n + 1)] for _ in range(m + 1)]

    for i in range(m + 1):
        dp[i][0] = i

    for j in range(n + 1):
        dp[0][j] = j

    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if s[i - 1] == t[j - 1]:
                dp[i][j] = dp[i - 1][j - 1]
            else:
                dp[i][j] = min(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1]) + 1

    return dp[m][n]


if __name__ == "__main__":
    print(edit_distance(input(), input()))
